flink-core/src/main/java/org/apache/flink/api/common/operators/base/OuterJoinOperatorBase.java

```diff
@@ -0,0 +1,314 @@
+
+			comparator = ((CompositeType<T>) typeInformation).createComparator(keyPositions, orders, 0, executionConfig);
+		} else {
+			throw new RuntimeException("Type information for input of type " + typeInformation.getClass()
+					.getCanonicalName() + " is not supported. Could not generate a comparator.");
+		}
+		return comparator;
+	}
+
+	private static class OuterJoinListIterator<IN1, IN2> {
+
+
+		private static enum MatchStatus {
+			NONE_REMAINED, FIRST_REMAINED, SECOND_REMAINED, FIRST_EMPTY, SECOND_EMPTY
+		}
+
+		private OuterJoinType outerJoinType;
+
+		private ListKeyGroupedIterator<IN1> leftGroupedIterator;
+		private ListKeyGroupedIterator<IN2> rightGroupedIterator;
+		private Iterable<IN1> currLeftSubset;
+		private ResettableIterator currLeftIterator;
+		private Iterable<IN2> currRightSubset;
+		private ResettableIterator currRightIterator;
+
+		private MatchStatus matchStatus;
+		private GenericPairComparator<IN1, IN2> pairComparator;
+
+		private IN1 leftReturn;
+		private IN2 rightReturn;
+
+		public OuterJoinListIterator(List<IN1> leftInput, TypeSerializer<IN1> leftSerializer, final TypeComparator<IN1> leftComparator,
+				List<IN2> rightInput, TypeSerializer<IN2> rightSerializer, final TypeComparator<IN2> rightComparator,
+				OuterJoinType outerJoinType) {
+			this.outerJoinType = outerJoinType;
+			pairComparator = new GenericPairComparator<>(leftComparator, rightComparator);
+			leftGroupedIterator = new ListKeyGroupedIterator<>(leftInput, leftSerializer, leftComparator);
+			rightGroupedIterator = new ListKeyGroupedIterator<>(rightInput, rightSerializer, rightComparator);
+			// ----------------------------------------------------------------
+			// Sort
+			// ----------------------------------------------------------------
+			Collections.sort(leftInput, new Comparator<IN1>() {
+				@Override
+				public int compare(IN1 o1, IN1 o2) {
+					return leftComparator.compare(o1, o2);
+				}
+			});
+
+			Collections.sort(rightInput, new Comparator<IN2>() {
+				@Override
+				public int compare(IN2 o1, IN2 o2) {
+					return rightComparator.compare(o1, o2);
+				}
+			});
+
+		}
+
+		@SuppressWarnings("unchecked")
+		private boolean next() throws IOException {
+			boolean hasMoreElements;
+			if ((currLeftIterator == null || !currLeftIterator.hasNext()) && (currRightIterator == null || !currRightIterator.hasNext())) {
+				hasMoreElements = nextGroups(outerJoinType);
+				if (hasMoreElements) {
+					if (outerJoinType != OuterJoinType.LEFT) {
+						currLeftIterator = new ListIteratorWrapper(currLeftSubset.iterator());
+					}
+					leftReturn = (IN1) currLeftIterator.next();
+					if (outerJoinType != OuterJoinType.RIGHT) {
+						currRightIterator = new ListIteratorWrapper(currRightSubset.iterator());
+					}
+					rightReturn = (IN2) currRightIterator.next();
+					return true;
+				} else {
+					//no more elements
+					return false;
+				}
+			} else if (currLeftIterator.hasNext() && !currRightIterator.hasNext()) {
```

Comment:
@jkovacs is right, it is not possible that ```currLeftIterator``` or ```currRightIterator``` are null at this point, as they are just wrappers around the subsets and always get assigned. In case that a subset is null and ```hasNext()``` is called, the wrapper will return false. 
There are also tests in ```OuterJoinOperatorBaseTest``` which cover the cases where one side of the outer join is empty.